tests/integration/test_jenkins.py

```diff
@@ -29,6 +32,94 @@ async def test_jenkins_update_ui_disabled(
+    ops_test: OpsTest,
+    model_app_unit: ModelAppUnit,
+    jenkins_version: str,
+    freeze_time: str,
+):
+    """
+    arrange: given jenkins charm with frozen time to 15:00 UTC.
+    act: when update-time-range between 3AM to 5AM is applied.
+    assert: the update does not take place.
+    """
+    await model_app_unit.model.wait_for_idle()
+    await model_app_unit.app.set_config({"update-time-range": "03-05"})
+    unit_name: str = model_app_unit.unit.name
+    charm_path = f"/var/lib/juju/agents/unit-{unit_name.replace('/', '-')}/charm"
+
+    # Overwrite requirements.txt
+    with tempfile.NamedTemporaryFile("wt", encoding="utf-8") as temp_requirements_txt, open(
+        "requirements.txt", encoding="utf-8"
+    ) as orig_requirements_txt:
+        requirements = "\n".join((orig_requirements_txt.read(), "freezegun"))
+        temp_requirements_txt.write(requirements)
+        temp_requirements_txt.flush()
+
+        await ops_test.juju(
+            "scp", temp_requirements_txt.name, f"{unit_name}:{charm_path}/requirements.txt"
+        )
+
+        # Install requirements.txt
+        await ops_test.juju(
+            "ssh",
+            unit_name,
+            "/usr/local/bin/pip",
+            "install",
+            "--target",
+            f"{charm_path}/venv",
+            "-r",
+            f"{charm_path}/requirements.txt",
+        )
+
+    # Overwrite charm.py
+    with tempfile.NamedTemporaryFile(mode="wt", encoding="utf-8") as temp_charm_py, open(
+        "src/charm.py", encoding="utf-8"
+    ) as orig_charm_py:
+        shebang_delimeter = "#!/usr/bin/env python3"
+        class_delimeter = "class JenkinsK8SOperatorCharm(CharmBase):"
+        charm_code = orig_charm_py.read()
+        split_code = charm_code.split(shebang_delimeter, 1)
+        split_code.insert(0, f"{shebang_delimeter}\nfrom freezegun import freeze_time\n")
+        charm_code = "".join(split_code)
+        split_code = charm_code.split(class_delimeter, 1)
+        split_code.insert(1, f'freeze_time("{freeze_time}").start()\n{class_delimeter}\n')
+        temp_charm_py.write("".join(split_code))
+        temp_charm_py.flush()
+
+        # unit.scp_to currently doesn't support k8s charms
+        # https://github.com/juju/python-libjuju/issues/885
+        (ret, _, stderr) = await ops_test.juju(
+            "scp", temp_charm_py.name, f"{unit_name}:{charm_path}/src/charm.py"
+        )
+        assert not ret, f"failed to scp modified charm, {stderr}"
+        (ret, _, stderr) = await ops_test.juju(
+            "ssh", unit_name, "/usr/bin/chmod", "755", f"{charm_path}/src/charm.py"
+        )
+        assert not ret, f"failed to chmod modified charm, {stderr}"
+        (ret, _, stderr) = await ops_test.juju(
+            "ssh", unit_name, "/usr/bin/chown", "root:root", f"{charm_path}/src/charm.py"
+        )
+        assert not ret, f"failed to chown modified charm, {stderr}"
```

Comment:
I'm not so fan of this way of doing. There's a library "libfaketime" that can normally be easily injected using LD_PRELOAD envvar
I'm trying to figure out if there's a way to pass any environment variable to the charm context that would enable this solution

Reply:
Ok I've 2 proposed approaches:
- Instead of trying to messup with the time, we can compute the time range based on the current hours (if it's 9AM and we want an interval that leads to an update then we do `f"{now().hour-1}-{now().hour+1}"` and if we want an interval that doesn't trigger an update `f"{now().hour+1}-{now().hour-1}"`
- I tested successfuly the following process:
```bash
juju run --unit traefik-k8s/0 -- apt update
juju run --unit traefik-k8s/0 -- apt install -y libfaketime
ubuntu@charm-dev:~$ juju run --unit traefik-k8s/0 -- 'LD_PRELOAD="/usr/lib/aarch64-linux-gnu/faketime/libfaketime.so.1" FAKETIME="@2000-01-01 11:12:13" date'
```
Which gives the following output:
`Sat Jan  1 11:12:13 UTC 2000`

Now if I run on a modified traefik charm that logs the current time:
```bash
juju run --unit traefik-k8s/0 -- 'LD_PRELOAD="/usr/lib/aarch64-linux-gnu/faketime/libfaketime.so.1" FAKETIME="@2000-01-01 11:12:13" JUJU_DISPATCH_PATH=hooks/upgrade-charm JUJU_MODEL_NAME=dqlite JUJU_UNIT_NAME=traefik-k8s/0 ./dispatch'
```

I'll get in the logs:
`unit-traefik-k8s-0: 17:47:32 INFO unit.traefik-k8s/0.juju-log 2000-01-01 11:12:14.219210`

Reply:
I think the environment variable solution would be a good thing to try here so that when other charms face similar issues, the current repository can be referred to for manipulating time. Thank you for your research & suggestion!

Reply:
Implemented to manually trigger update-status-hook using libfaketime environment variables.